tests/protocol/mocks.py

```diff
@@ -19,124 +19,169 @@
+            handler = http_handlers.put
 
         if handler is not None:
-            return_value = handler(url, *args, **kwargs)
+            if method == 'GET':
+                return_value = handler(url, **kwargs)
+            else:
+                return_value = handler(url, data, **kwargs)
             if return_value is not None:
                 tracked_urls.append(url)
                 if isinstance(return_value, Exception):
                     raise return_value
                 return return_value
 
-        return original_http_request(method, url, *args, **kwargs)
+        # if the request was not handled try to use the mock wireserver data
+        if _USE_MOCK_WIRE_DATA_RE.match(url) is not None:
+            if method == 'GET':
+                return protocol.mock_wire_data.mock_http_get(url, **kwargs)
+            if method == 'POST':
+                return protocol.mock_wire_data.mock_http_post(url, data, **kwargs)
+
+        # the request was not handled; fail or call the original resutil.http_request
+        if fail_on_unknown_request:
+            raise ValueError('Unknown HTTP request: {0} [{1}]'.format(url, method))
+        return original_http_request(method, url, data, **kwargs)
+
+    #
+    # functions to start/stop the mocks
+    #
+    def start():
+        patched = patch("azurelinuxagent.common.utils.restutil.http_request", side_effect=http_request)
+        patched.start()
+        start.http_request_patch = patched
+
+        patched = patch("azurelinuxagent.common.protocol.wire.CryptUtil", side_effect=protocol.mock_wire_data.mock_crypt_util)
+        patched.start()
+        start.crypt_util_patch = patched
+    start.http_request_patch = None
+    start.crypt_util_patch = None
 
-    patched = patch("azurelinuxagent.common.utils.restutil.http_request", side_effect=http_request)
-    patched.track_url = lambda url: tracked_urls.append(url)
-    patched.get_tracked_urls = lambda: tracked_urls
-    patched.start()
+    def stop():
+        if start.crypt_util_patch is not None:
+            start.crypt_util_patch.stop()
+        if start.http_request_patch is not None:
+            start.http_request_patch.stop()
+
+    #
+    # create the protocol object
+    #
+    protocol = WireProtocol(restutil.KNOWN_WIRESERVER_IP)
+    protocol.mock_wire_data = mockwiredata.WireProtocolData(mock_wire_data_file)
+    protocol.start = start
+    protocol.stop = stop
+    protocol.track_url = lambda url: tracked_urls.append(url)
+    protocol.get_tracked_urls = lambda: tracked_urls
+    protocol.set_http_handlers = lambda http_get_handler=None, http_post_handler=None, http_put_handler=None:\
+        http_handlers(get=http_get_handler, post=http_post_handler, put=http_put_handler)
+
+    # go do it
     try:
-        yield patched
+        protocol.start()
+        protocol.detect()
+        yield protocol
     finally:
-        patched.stop()
+        protocol.stop()
+
+
+class HttpRequestPredicates(object):
```

Comment:
The tests were doing those checks over and over (and I anticipate more of those as I continue refactoring the protocol) so I captured them here. Also, improved the checks to also test for the wireserver endpoint address instead of just looking at substrings in the urls